SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -0,0 +1,144 @@
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Schema/AuthorityIntent.h"
+#include "SpatialCommonTypes.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialLoadBalanceEnforcer);
+
+using namespace SpatialGDK;
+
+USpatialLoadBalanceEnforcer::USpatialLoadBalanceEnforcer(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+	, NetDriver(nullptr)
+	, VirtualWorkerTranslator(nullptr)
+{
+}
+
+void USpatialLoadBalanceEnforcer::Init(USpatialNetDriver* InNetDriver, SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator)
+{
+	NetDriver = InNetDriver;
+	VirtualWorkerTranslator = InVirtualWorkerTranslator;
+}
+
+void USpatialLoadBalanceEnforcer::Tick()
+{
+	ProcessQueuedAclAssignmentRequests();
+}
+
+void USpatialLoadBalanceEnforcer::OnComponentUpdated(const Worker_ComponentUpdateOp& Op)
+{
+	check(NetDriver)
+	check(NetDriver->StaticComponentView)
+	if (Op.update.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID &&
+		NetDriver->StaticComponentView->GetAuthority(Op.entity_id, SpatialConstants::ENTITY_ACL_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE)
+	{
+		QueueAclAssignmentRequest(Op.entity_id);
+	}
+}
+
+void USpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+{
+	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
+		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+	{
+		// We have gained authority over the ACL component for some entity
+		// If we have authority, the responsibility here is to set the EntityACL write auth to match the worker requested via the virtual worker component
+		QueueAclAssignmentRequest(AuthOp.entity_id);
+	}
+}
+
+void USpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
+{
+	if (AclWriteAuthAssignmentRequests.ContainsByPredicate([EntityId](const WriteAuthAssignmentRequest& Request) { return Request.EntityId == EntityId; }))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("(%s) Already has an ACL authority request for entity %lld"), *NetDriver->Connection->GetWorkerId(), EntityId);
+	}
+	else
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("(%s) Queueing ACL assignment request for %lld"), *NetDriver->Connection->GetWorkerId(), EntityId);
+		AclWriteAuthAssignmentRequests.Add(WriteAuthAssignmentRequest(EntityId));
+	}
+}
+
+void USpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
+{
+	TArray<Worker_EntityId> CompletedRequests;
+	CompletedRequests.Reserve(AclWriteAuthAssignmentRequests.Num());
+
+	for (WriteAuthAssignmentRequest& Request : AclWriteAuthAssignmentRequests)
+	{
+		static const int16_t ConcerningNumAttmempts = 5;
+		if (Request.ProcessAttempts >= ConcerningNumAttmempts)
+		{
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Failed to process WriteAuthAssignmentRequest with EntityID: %lld. Process attempts made: %d"), Request.EntityId, Request.ProcessAttempts);
```

Comment:
Should we `continue` at this point?

Reply:
I think we should still try and process it. This is more to log whether we ever have a situation where for some reason we can not send the updates.